exec/local.go

```diff
@@ -109,34 +140,39 @@ func (l *localExecutor) Run(task *Task) {
 			in = append(in, reader)
 		}
 	}
-	task.Set(TaskRunning)
-
-	// Start execution, then place output in a task buffer. We also plumb a
-	// metrics scope in here so we can store and aggregate metrics.
-	task.Scope.Reset(nil)
-	out := task.Do(in)
-	buf, err := bufferOutput(metrics.ScopedContext(ctx, &task.Scope), task, out)
-	task.Lock()
-	if err == nil {
-		l.mu.Lock()
-		l.buffers[task] = buf
-		l.mu.Unlock()
-		task.state = TaskOk
-	} else {
-		task.state = TaskErr
-		task.err = err
-	}
-	task.Broadcast()
-	task.Unlock()
+	return in, nil
 }
 
 func (l *localExecutor) Reader(task *Task, partition int) sliceio.ReadCloser {
 	l.mu.Lock()
-	buf := l.buffers[task]
+	buf, ok := l.buffers[task]
 	l.mu.Unlock()
+	if !ok {
+		return sliceio.ReaderWithCloseFunc{
+			Reader:    sliceio.ErrReader(fmt.Errorf("no data for %v", task)),
+			CloseFunc: func() error { return nil },
+		}
+	}
 	return buf.Reader(partition)
 }
 
+func (l *localExecutor) Discard(_ context.Context, task *Task) {
+	if !task.Combiner.IsNil() && task.CombineKey != "" {
+		// We do not yet handle tasks with shared combiners.
+		return
+	}
+	task.Lock()
+	if task.state == TaskOk {
+		l.mu.Lock()
+		delete(l.buffers, task)
+		l.mu.Unlock()
+		task.Unlock()
+		task.Set(TaskLost)
```

Comment:
Should this assign to `task.state` directly before holding the lock, to avoid a race?

Reply:
Yes. Fixed.